File: c_komega/src/komega_parameter.c
/**
 * @file komega_parameter.c
 * @brief Parameter module for Komega C library
 * 
 * This module provides global parameter management for the Komega library.
 * It corresponds to the komega_parameter Fortran module and handles
 * solver parameters, iteration counts, and convergence thresholds.
 * 
 * @author Mitsuaki Kawamura (Original Fortran)
 * @author C Port for verification and testing purposes
 * @date 2025
 * @version 1.0.0
 * 
 * @see komega_parameter.h
 * @see komega.h
 */

#include "komega_parameter.h"
#include <stdlib.h>
#include <string.h>

/* Global parameter instance */
static komega_parameter_t* global_params = NULL;

/**
 * @brief Create a new parameter instance
 * 
 * Allocates memory for a new parameter structure and initializes it
 * with default values. The default values are:
 * - ndim = 0
 * - nl = 0  
 * - nz = 0
 * - itermax = 100
 * - threshold = 1e-6
 * - almost0 = KOMEGA_ALMOST_ZERO
 * - iter = 0
 * - resnorm = 0.0
 * - initialized = false
 * 
 * @return Pointer to new parameter instance, or NULL on failure
 * 
 * @see komega_parameter_destroy()
 * @see komega_parameter_initialize()
 */
komega_parameter_t* komega_parameter_create(void) {
    komega_parameter_t* params = malloc(sizeof(komega_parameter_t));
    if (params == NULL) {
        return NULL;
    }
    
    /* Initialize with default values */
    params->ndim = 0;
    params->nl = 0;
    params->nz = 0;
    params->itermax = 100;
    params->threshold = 1e-6;
    params->almost0 = KOMEGA_ALMOST_ZERO;
    params->iter = 0;
    params->resnorm = 0.0;
    params->initialized = false;
    
    return params;
}

/**
 * @brief Destroy a parameter instance
 * 
 * Frees the memory allocated for the parameter structure.
 * Safe to call with NULL pointer.
 * 
 * @param params Pointer to parameter instance to destroy
 * 
 * @see komega_parameter_create()
 */
void komega_parameter_destroy(komega_parameter_t* params) {
    if (params != NULL) {
        free(params);
    }
}

/**
 * @brief Initialize parameter instance with solver parameters
 * 
 * Sets up the parameter structure with the given solver parameters.
 * All parameters are validated before setting. The function will
 * return an error if any parameter is invalid (<= 0).
 * 
 * @param params Pointer to parameter instance
 * @param ndim Dimension of the problem (must be > 0)
 * @param nl Local dimension (must be > 0)
 * @param nz Number of frequencies (must be > 0)
 * @param itermax Maximum number of iterations (must be > 0)
 * @param threshold Convergence threshold (must be > 0)
 * 
 * @return KOMEGA_SUCCESS on success, KOMEGA_ERROR_INVALID_PARAM on failure
 * 
 * @see komega_parameter_create()
 */
komega_status_t komega_parameter_initialize(komega_parameter_t* params, 
                                           int ndim, int nl, int nz, 
                                           int itermax, double threshold) {
    if (params == NULL) {
        return KOMEGA_ERROR_INVALID_PARAM;
    }
    
    if (ndim <= 0 || nl <= 0 || nz <= 0 || itermax <= 0 || threshold <= 0.0) {
        return KOMEGA_ERROR_INVALID_PARAM;
    }
    
    params->ndim = ndim;
    params->nl = nl;
    params->nz = nz;
    params->itermax = itermax;
    params->threshold = threshold;
    params->iter = 0;
    params->resnorm = 0.0;
    params->initialized = true;
    
    return KOMEGA_SUCCESS;
}

/**
 * @brief Increment the iteration counter
 * 
 * Increments the current iteration count by 1.
 * Safe to call with NULL pointer.
 * 
 * @param params Pointer to parameter instance
 */
void komega_parameter_increment_iteration(komega_parameter_t* params) {
    if (params != NULL) {
        params->iter++;
    }
}

/**
 * @brief Reset the iteration counter
 * 
 * Resets the current iteration count to 0.
 * Safe to call with NULL pointer.
 * 
 * @param params Pointer to parameter instance
 */
void komega_parameter_reset_iteration(komega_parameter_t* params) {
    if (params != NULL) {
        params->iter = 0;
    }
}

/**
 * @brief Check if parameters are initialized
 * 
 * @param params Pointer to parameter instance
 * @return true if initialized, false otherwise
 */
bool komega_parameter_is_initialized(const komega_parameter_t* params) {
    return (params != NULL) ? params->initialized : false;
}

/**
 * @brief Get the problem dimension
 * 
 * @param params Pointer to parameter instance
 * @return Problem dimension (ndim), or 0 if params is NULL
 */
int komega_parameter_get_ndim(const komega_parameter_t* params) {
    return (params != NULL) ? params->ndim : 0;
}

/**
 * @brief Get the local dimension
 * 
 * @param params Pointer to parameter instance
 * @return Local dimension (nl), or 0 if params is NULL
 */
int komega_parameter_get_nl(const komega_parameter_t* params) {
    return (params != NULL) ? params->nl : 0;
}

/**
 * @brief Get the number of frequencies
 * 
 * @param params Pointer to parameter instance
 * @return Number of frequencies (nz), or 0 if params is NULL
 */
int komega_parameter_get_nz(const komega_parameter_t* params) {
    return (params != NULL) ? params->nz : 0;
}

/**
 * @brief Get the maximum number of iterations
 * 
 * @param params Pointer to parameter instance
 * @return Maximum number of iterations (itermax), or 0 if params is NULL
 */
int komega_parameter_get_itermax(const komega_parameter_t* params) {
    return (params != NULL) ? params->itermax : 0;
}

/**
 * @brief Get the convergence threshold
 * 
 * @param params Pointer to parameter instance
 * @return Convergence threshold, or 0.0 if params is NULL
 */
double komega_parameter_get_threshold(const komega_parameter_t* params) {
    return (params != NULL) ? params->threshold : 0.0;
}

/**
 * @brief Get the current iteration count
 * 
 * @param params Pointer to parameter instance
 * @return Current iteration count, or 0 if params is NULL
 */
int komega_parameter_get_iter(const komega_parameter_t* params) {
    return (params != NULL) ? params->iter : 0;
}

/**
 * @brief Get the residual norm
 * 
 * @param params Pointer to parameter instance
 * @return Current residual norm, or 0.0 if params is NULL
 */
double komega_parameter_get_resnorm(const komega_parameter_t* params) {
    return (params != NULL) ? params->resnorm : 0.0;
}

/**
 * @brief Set the residual norm
 * 
 * @param params Pointer to parameter instance
 * @param resnorm New residual norm value
 */
void komega_parameter_set_resnorm(komega_parameter_t* params, double resnorm) {
    if (params != NULL) {
        params->resnorm = resnorm;
    }
}

/**
 * @brief Get the global parameter instance
 * 
 * Returns a singleton instance of the parameter structure.
 * Creates a new instance if none exists.
 * 
 * @return Pointer to global parameter instance
 * 
 * @see komega_cleanup_global_parameter()
 */
komega_parameter_t* komega_get_global_parameter(void) {
    if (global_params == NULL) {
        global_params = komega_parameter_create();
    }
    return global_params;
}

/**
 * @brief Clean up the global parameter instance
 * 
 * Destroys the global parameter instance and frees its memory.
 * 
 * @see komega_get_global_parameter()
 */
void komega_cleanup_global_parameter(void) {
    if (global_params != NULL) {
        komega_parameter_destroy(global_params);
        global_params = NULL;
    }
}

File: python_komega/komega_parameter.py
"""
Komega Parameter Module - Global parameters and constants

This module corresponds to the Fortran komega_parameter module and contains
global parameters used throughout the Komega library.

Copyright (C) 2016 Mitsuaki Kawamura
Python port created for verification and testing purposes.
"""

import numpy as np
from typing import Optional, List


class KomegaParameter:
    """
    Global parameters for the Komega library.
    
    This class manages global parameters that are used across different
    solver modules in the Komega library.
    """
    
    def __init__(self):
        """Initialize global parameters with default values."""
        # Mathematical constants
        self.almost0 = 1e-50
        
        # MPI and communication parameters
        self.comm: Optional[int] = None
        self.lmpi: bool = False
        
        # Problem dimensions
        self.ndim: int = 0  # Dimension of Hamiltonian
        self.nl: int = 0    # Dimension of projection
        self.nz: int = 0    # Number of frequencies (shifts)
        
        # Iteration parameters
        self.itermax: int = 0  # Maximum number of iterations
        self.iter: int = 0     # Current iteration counter
        self.iz_seed: int = 0  # Index of frequency seed
        
        # Convergence parameters
        self.threshold: float = 1e-6  # Convergence threshold
        self.resnorm: float = 0.0     # Residual norm
        
        # Convergence flags for each frequency
        self.lz_conv: Optional[List[bool]] = None
        
    def initialize(self, ndim: int, nl: int, nz: int, itermax: int, 
                   threshold: float, comm: Optional[int] = None) -> None:
        """
        Initialize parameters for a new problem.
        
        Parameters
        ----------
        ndim : int
            Dimension of the Hamiltonian matrix
        nl : int
            Dimension of the projection space
        nz : int
            Number of frequency points
        itermax : int
            Maximum number of iterations
        threshold : float
            Convergence threshold
        comm : int, optional
            MPI communicator (for MPI version)
        """
        self.ndim = ndim
        self.nl = nl
        self.nz = nz
        self.itermax = itermax
        self.threshold = threshold
        self.iter = 0
        self.iz_seed = 1
        self.resnorm = 0.0
        
        # Initialize convergence flags
        self.lz_conv = [False] * nz
        
        # MPI setup
        if comm is not None:
            self.comm = comm
            self.lmpi = True
        else:
            self.comm = None
            self.lmpi = False
    
    def reset_iteration(self) -> None:
        """Reset iteration counter."""
        self.iter = 0
    
    def increment_iteration(self) -> None:
        """Increment iteration counter."""
        self.iter += 1
    
    def set_seed_frequency(self, iz: int) -> None:
        """
        Set the seed frequency index.
        
        Parameters
        ----------
        iz : int
            Index of the seed frequency
        """
        self.iz_seed = iz
    
    def check_convergence(self, residual_norms: np.ndarray) -> bool:
        """
        Check convergence for all frequencies.
        
        Parameters
        ----------
        residual_norms : np.ndarray
            Residual norms for each frequency
            
        Returns
        -------
        bool
            True if all frequencies have converged
        """
        if self.lz_conv is None:
            return False
            
        # Update convergence flags
        for i in range(self.nz):
            if not self.lz_conv[i] and residual_norms[i] < self.threshold:
                self.lz_conv[i] = True
        
        # Check if all frequencies have converged
        return all(self.lz_conv)
    
    def get_converged_frequencies(self) -> List[int]:
        """
        Get list of converged frequency indices.
        
        Returns
        -------
        List[int]
            List of converged frequency indices
        """
        if self.lz_conv is None:
            return []
        return [i for i, converged in enumerate(self.lz_conv) if converged]
    
    def get_unconverged_frequencies(self) -> List[int]:
        """
        Get list of unconverged frequency indices.
        
        Returns
        -------
        List[int]
            List of unconverged frequency indices
        """
        if self.lz_conv is None:
            return list(range(self.nz))
        return [i for i, converged in enumerate(self.lz_conv) if not converged]


# Global parameter instance
_global_params = KomegaParameter()


def get_global_params() -> KomegaParameter:
    """
    Get the global parameter instance.
    
    Returns
    -------
    KomegaParameter
        Global parameter instance
    """
    return _global_params


def initialize_global_params(ndim: int, nl: int, nz: int, itermax: int, 
                           threshold: float, comm: Optional[int] = None) -> None:
    """
    Initialize global parameters.
    
    Parameters
    ----------
    ndim : int
        Dimension of the Hamiltonian matrix
    nl : int
        Dimension of the projection space
    nz : int
        Number of frequency points
    itermax : int
        Maximum number of iterations
    threshold : float
        Convergence threshold
    comm : int, optional
        MPI communicator (for MPI version)
    """
    _global_params.initialize(ndim, nl, nz, itermax, threshold, comm)


def reset_global_params() -> None:
    """Reset global parameters to default values."""
    global _global_params
    _global_params = KomegaParameter()
